src/core/geometry/qgsgeometryeditutils.cpp

```diff
@@ -131,28 +131,22 @@ Qgis::GeometryOperationResult QgsGeometryEditUtils::addPart( QgsAbstractGeometry
       {
         poly = std::make_unique< QgsCurvePolygon >();
       }
-      // Ownership is still with part, curve points to the same object and is transferred
-      // to poly here.
-      part.release();
-      poly->setExteriorRing( curve );
+      poly->setExteriorRing( qgsgeometry_cast<QgsCurve *>( part.release() ) );
       added = geomCollection->addGeometry( poly.release() );
     }
     else if ( QgsWkbTypes::flatType( part->wkbType() ) == QgsWkbTypes::Polygon
               || QgsWkbTypes::flatType( part->wkbType() ) == QgsWkbTypes::Triangle
               || QgsWkbTypes::flatType( part->wkbType() ) == QgsWkbTypes::CurvePolygon )
     {
-      QgsCurvePolygon *curvePolygon = qgsgeometry_cast<QgsCurvePolygon *>( part.get() );
-      if ( curvePolygon )
+      if ( const QgsCurvePolygon *curvePolygon = qgsgeometry_cast< const QgsCurvePolygon *>( part.get() ) )
```

Comment:
is there any reason to use this rather than the 2 former lines.
I find this a bit more readable/explicit.

Reply:
Yes -- we should limit the scope of curvePolygon to only inside the if block, as otherwise it's a pointer to a deleted object which is never cleared. Plus it's just general good practice to limit scope of variables to as tight as possible...

Reply:
Ok, thanks for the explanation